ui/cockpit/src/modules/utils/notifications.js

```diff
@@ -0,0 +1,154 @@
+}
+
+/**
+ *
+ * Notification object may specify the following fields:
+ *   type: type of the notification (info, warning, danger, success)
+ *   status: main status line
+ *   message: detail message
+ *   unsafe: boolean, indicates that status and message should not be sanitized
+ *   duration: time duration in ms the notification should be shown to the user
+ *   exclusive: boolean || array of attribute names that notification should be exclusive with
+ *   scope: notification will be removed when the specified scope is destroyed
+ *
+ * @param notification {notification}
+ * @returns {undefined}
+ */
+
+export function addError(error) {
+  if (!error.type) {
+    error.type = "danger";
+  }
+
+  add(error);
+}
+
+export function addMessage(message) {
+  if (!message.type) {
+    message.type = "info";
+  }
+
+  add(message);
+}
+
+export function add(notification) {
+  const exclusive = notification.exclusive;
+
+  if (!notification.unsafe) {
+    notification.status = escapeHtml(notification.status);
+    notification.message = escapeHtml(notification.message || "");
+  }
+
+  if (exclusive) {
+    if (typeof exclusive == "boolean") {
+      clearAll();
+    } else {
+      var filter = {};
+      exclusive.forEach(key => {
+        filter[key] = notification[key];
+      });
+
+      clear(filter);
+    }
+  }
+
+  notifications.push(notification);
+
+  for (var i = consumers.length - 1, c; (c = consumers[i]); i--) {
+    // add to first interested consumer only
+    if (c.add(notification)) {
+      break;
+    }
+  }
+
+  if (notification.duration) {
+    window.setTimeout(function() {
+      if (notification.scope) {
+        delete notification.scope;
+      }
+      clear(notification);
+    }, notification.duration);
+  }
+
+  if (notification.scope) {
+    notification.scope.$on("$destroy", function() {
+      // remove the scope from the notification object to resolve circular dependency
+      // when clearing the notification
+      delete notification.scope;
+      clear(notification);
+    });
+  }
+}
+
+export function clear(notification) {
+  var removeCandidates = [];
+
+  if (typeof notification == "string") {
+    notification = { status: notification };
+    removeCandidates = notifications.filter(
+      candidate => candidate.status === notification.status
+    );
+  }
+
+  removeCandidates.push(notification);
+
+  removeCandidates.forEach(e => {
+    var idx = notifications.indexOf(e);
+    if (idx !== -1) {
+      notifications.splice(idx, 1);
+    }
+
+    consumers.forEach(consumer => {
+      consumer.remove(e);
+    });
+  });
+}
+
+export function clearAll() {
+  while (notifications.length) {
+    var notification = notifications.pop();
+    clear(notification);
+  }
+}
+
+export function registerConsumer(consumer, index) {
+  if (typeof index === "number" && consumers.length > index) {
+    consumers.splice(index, 0, consumer);
+    return index;
+  } else {
+    return consumers.push(consumer) - 1;
+  }
+}
+
```

Comment:
Adjusted to insert React Components at the same position than before on re-render. The rest is the same as the angular service, minus the removed Angular services.